src/OrderedMsgChain.js

```diff
@@ -0,0 +1,108 @@
+import debugFactory from 'debug'
+import Heap from 'heap'
+import { Errors, MessageLayer } from 'streamr-client-protocol'
+
+const debug = debugFactory('StreamrClient::OrderedMsgChain')
+const { MessageRef } = MessageLayer
+
+const DEFAULT_GAPFILL_TIMEOUT = 5000
+const MAX_GAP_REQUESTS = 10
+
+export default class OrderedMsgChain {
+    constructor(publisherId, msgChainId, inOrderHandler, gapHandler, gapFillTimeout = DEFAULT_GAPFILL_TIMEOUT) {
+        this.publisherId = publisherId
+        this.msgChainId = msgChainId
+        this.inOrderHandler = inOrderHandler
+        this.gapHandler = gapHandler
+        this.lastReceivedMsgRef = null
+        this.gapFillTimeout = gapFillTimeout
+        this.queue = new Heap((msg1, msg2) => {
+            return msg1.getMessageRef().compareTo(msg2.getMessageRef())
+        })
+    }
+
+    add(unorderedStreamMessage) {
+        const msgRef = unorderedStreamMessage.getMessageRef()
+        if (this.lastReceivedMsgRef && msgRef.compareTo(this.lastReceivedMsgRef) < 0) {
+            // Prevent double-processing of messages for any reason
+            debug('Already received message: %o, lastReceivedMsgRef: %d. Ignoring message.', msgRef, this.lastReceivedMsgRef)
+        }
+
+        if (this._isNextMessage(unorderedStreamMessage)) {
+            this._process(unorderedStreamMessage)
+        } else {
+            if (!this.gap) {
```

Comment:
Is it possible for `setInterval` to return a falsey value? Changing the comparison to `this.gap != null` would unburden the reader from that thought